IPIP/0000-gateway-json-cbor-response-format.md

```diff
@@ -0,0 +1,79 @@
+# IPIP 0000: Gateway JSON and TAR Response Formats
+
+- Start Date: 2022-10-07
+- Related Issues:
+  - [ipfs/in-web-browsers/issues/182]
+  - [ipfs/specs/pull/328]
+  - [ipfs/kubo/issues/8823]
+  - [ipfs/kubo/pull/9335]
+  - [ipfs/go-ipfs/issues/7552]
+
+## Summary
+
+Add [DAG-JSON], [DAG-CBOR], JSON and CBOR response formats to the [HTTP Gateway](../http-gateways/).
+
+## Motivation
+
+Currently, the gateway supports requesting data in the [DAG-PB], RAW, and [CAR]
+formats. However, it should be possible to download everything over the gateway
+in order to unlock the potentials of the [IPLD Data Model][ipld-data-model].
+
+For example, DAG-PB has a [logical format][dag-pb-format] which makes it possible
+to represent a DAG-PB directory as a [DAG-JSON] document. This means that, if we
+support DAG-JSON in the gateway, then we would support
+[JSON responses for directory listings][ipfs/go-ipfs/issues/7552], which has been
+requested by our users in the past.
+
+In addition, this functionality is already present on the current Kubo CLI. By
+bringing it to the gateways, we provide users with more power when it comes
+to traversing IPLD data.
+
+## Detailed design
+
+The solution is to allow the Gateway to support serializing an IPLD model
+representation as [DAG-JSON], [DAG-CBOR], JSON and CBOR by requesting them
+using either the `Accept` HTTP header or the `format` URL query.
+
+## Test fixtures
+
+TODO
```

Comment:
@lidel Unfortunately, I'm not sure if this will be easy to do. See https://github.com/ipfs/kubo/pull/9335#issuecomment-1277057863. It seems that the Go codecs encoders are quite strict, making the differences between both codecs very minimal. I thought about playing with invalid links, such as:

```
{ "foo": { "/": "this is no cid" } }
```

But, if I add it as json, it works. But it also works to fetch it back as dag-json 😅 Adding it as dag-json, however, does not work.

Reply:
Test fixtures are also welcome at https://github.com/ipld/codec-fixtures :) We especially are missing failing ones. So if you create some, perhaps also add them there.

Reply:
@vmx @lidel regarding invalid versions, I'm thinking about what we could do. We would need something that is valid in some format, such as CBOR, but not valid in other format, such as JSON. The problem being that we would need to add the data to IPFS somehow and the serialisation process would already check if the data is valid or not for the codec.

Considering that DAG-JSON and DAG-CBOR are both subsets of their non DAG-* versions, I can't think of anything on top of my mind that would work. We could see if we could add a weird CBOR that would fail when requested via `format=json`/`format=dag-json` (as both are equivalent when requesting a CBOR document.

I think this is related to what I mentioned on the previous comment: that the Go impl. of CBOR and JSON are quite strict by themselves, so adding something to IPFS that would then fail when downloading through the gateway would be quite hard. Am I missing anything?

Reply:
I have to admit I'm not deep into this whole IPIP, I just read "fixtures" and "DAG-CBOR/JSON" and jumped in :)

Valid JSON, but invalid DAG-JSON is anything with strange things after a property with just a slash, e.g. `{"/": true}`. Valid CBOR, but invalid DAG-CBOR: tags that are not tag 42. But of course all valid DAG-CBOR/JSON should be valid in the non DAG*-versions. Also valid DAG-JSON and DAG-CBOR should be interchangeable with each other.